### ISSUE TITLE
[1.3][proposal] import from NPM on local package

### ISSUE THREAD

**Issue by bySabi** [meteor/meteor#5890]
From a local package I can´t `import` or `require` from installed npm's on by  project `node_modules`
I had to explicit declare:
 `Npm.depends({'npm': 'version'})` inside `package.js`

With this behaviour meteor packages can depend only on published npm´s. Can´t depend on local indevelopment npm´s.

Is more suitable reuse application wide `node_modules` than package related `.npm` folder.

**Comment by trusktr** [comment]
I think this is a bug, and possibly a duplicate of #5866.

**Comment by bySabi** [comment]
I don´t think is a bug, yet. Only that developers ,@benjamn @avital don't arrive to this point, yet. 
Meteor rebuild npm's package "depends" on any package change and this installed npm don´t follow the same format than new 1.3 `node_modules` npm.
That's why I think that `Npm.depends` work like do on 1.2 and is not aware of new 1.3 `node_modules`
That's why I think is not related to: https://github.com/meteor/meteor/issues/5866

**Comment by trusktr** [comment]
Oh okay, yeah, Npm.depends works the old way. I don't think they've considered making it look in local `node_modules`, though that might be a good idea!

**Comment by bySabi** [comment]
"backward compatibility" is a must of course but ...

`packages` should be aware of `node_modules`, maybe improving `Npm.depends` and embrance npm 3 flattened dependencies. Not doing it will become `packages` based projects obsoleted.

**Comment by trusktr** [comment]
That could be true. I'm not sure, but part of the problem might be that `Npm.depends` uses NPM 2.x, while in Meteor 1.3 the user might use NPM 3.x (the version on NPM isn't enforced, you can use NPM 2.x if you want).

In theory, the `require` mechanism of Node.js is agnostic to the version of NPM used, so it should just work though.

**Comment by bySabi** [comment]
Maybe one solution for backard compatibility is relay on existing behaviour when `Npm.depends` it´s explicitly declared on `package.js` and ...
 search on `node_modules`, new 1.3 behaviour,  when `Npm.depends` is not declared.

Let´s see what MDG propose @benjamn @avital @stubailo

**Comment by timbotnik** [comment]
It seems we should address this in 1.3, as there are a number of GH issues related to local / in-development package support.  #5891 #5070 #2876

**Comment by bySabi** [comment]
Hi @benjamn I was reading your words on NPM and local Meteor package:  https://forums.meteor.com/t/meteor-1-3-early-beta-now-available/14723/122, and you raise a similiar approach that I propose here. And it is ...
For `meteor` packages local and 'remote':
1- `Npm.depends` on `package.js` remains for backwards compatibility with meteor 1.2.
2- Not explicitly declared `Npm.depends` will force meteor to search npm´s on per application `node_modules`

I was thinking about this proposal and it has a problem... 
We still need declare what npm require a package. 
Options are:
1- Reimplement `Npm.depends` with a new behaviour. For backward compatibility relay on `api.versionsFrom`. This can be really confusing.
2- add a `package.json` per meteor package like a normal npm. This is worse, two different package format together !!!??
3- add a new API, say: `Npm.dependecies`. I thing this is a easy way to manage the problem.

**Comment by benjamn** [comment]
What if you could do `api.imply(packageName)`, where `packageName` is the name of an npm package specified by `Npm.depends`? In other words, Meteor packages could imply not only other Meteor packages, but also npm packages.

If we go in that direction, we should probably allow `Npm.depends` to specify non-exact package versions (for example, `Npm.depends({ react: "^0.4" })`), so that multiple Meteor packages can imply compatible versions of the same npm package.

**Comment by trusktr** [comment]
> `Npm.depends({ react: "^0.14" })`

That sounds good. What would the user need to do with regards to `package.json` in the app root and those modules that will be manually installed manually via `npm install`? Will Meteor package NPM dependencies be handled separately?

**Comment by bySabi** [comment]
Something like this looks weired ??
```
...
  api.imply([
    {react: "^0.14"},        //  a npm react
    'react-tab@0.15.0'       //  a meteor package
  ], client);
...
```
The above someday I might like.


@benjamn you mean this, right??
```
...
Npm.depends({ react: "^0.14" });
...
Package.onUse(function(api) {
   api.imply([
      'react',            //  I'm a npm package. Look up!. 
      'react-tab@0.15.0'  //  I'm a lifelong meteor package. Don´t you see my tail (@0.15.0)
   ], 'client');
});
...
```

But this don´t solve the problem of `Npm.depends`,  `package\.npm\node_modules` vs `app\node_modules` 

If we do this for npm packages ??

```
...
  api.use([
    {react: "^0.14"},        //  a npm react
    'react-tab@0.15.0'       //  a meteor package
  ], client);
...
```

**Comment by trusktr** [comment]
I think the package API might be getting too confusing with these additions.

What if a package can say something like `api.useModules()`, then the rest of the `package.js` API would remain the same.

If `api.useModules()` appears in a `package.js` file, then Meteor will expect the Meteor package to contain a `package.json` with the NPM dependencies that the Meteor package needs.

At this point the developer of the Meteor has options (any or all can be chosen):

1. Don't export anything in `package.js`, let the user potentially `import` files from the Meteor package.
2. Make any number of `main.js` files if there is some logic to be automatically executed. If there are no `main.js` files, nothing will load unless `import`ed by the end user as per the previous option.
3. Export things in `package.js`. These things must be exported from, or be globals in, any of the Meteor package's `main.js` files.

Badaboom, easy.

**Comment by trusktr** [comment]
This makes it easy to separate the two concerns: NPM style behavior and Meteor package behavior.

The name of `api.useModules` is up for debate.

**Comment by bySabi** [comment]
I like this approach ...
```
...
  api.use([
    {                        // npm packages
      "react": "^0.14",
      "react-redux": "^0.1.1"
    },
    'react-tab@0.15.0'       // a meteor package
  ], client);
...
```
... because:

Make npm a first-class citizen. Just `use` a npm like a meteor package on `server` or `client` without fear.

I don´t like `api.useModules()` cause:
1- is just a rename of `Npm.depends`
2- is confuse semantically: What is a `package` and what is a `module`? who is who? 
3- require a `package.json` is add unneeded complexity to something that must be simple at start.

I don't close the door to `.json` settings files like do for example `cosmos:browserify` but this extra settings are unneeded for use most npm´s today.

**Comment by bySabi** [comment]
We need some feedback from someone else involved on meteor npm integration:
cc @arunoda @elidoran
a some others more:
cc @awatson1978 @SachaG @AdamBrodzinski

**Comment by trusktr** [comment]
> is just a rename of Npm.depends

No, it's not a rename. `api.useModule()` doesn't take any arguments.

```js
api.useModule({...}) // error
```

It only tells Meteor that the package is being treated like an NPM package.

> is confuse semantically: What is a package and what is a module? who is who? 

That's why I mentioned the name is up for debate. Maybe `api.usesNpmModules()` is better (with no arguments)?

> require a package.json is add unneeded complexity to something that must be simple at start.

I disagree. This is what the whole rest of the JavaScript community already does, and by adding 

```js
...
  api.use([
    {                        // npm packages
      "react": "^0.14",
      "react-redux": "^0.1.1"
    },
    'react-tab@0.15.0'       // a meteor package
  ], client);
...
```

then now we are having to know two ways of dealing with NPM packages: `package.json` in the app, and that `api.use` in Meteor packages.

I think it makes sense to keep the current Meteor package API as unchanged as possible, then introduce the existing NPM workflows that everyone already knows.

**Comment by bySabi** [comment]
@trusktr I agree with you about using `package.json` on app and `package.js with npm` on packages can be cofusing. But still we need a way to `define` npm modules/packages for import it in a isomorphic way, say on `client`, `server`, or both. I doubt that MDG relay on `package.json` for this cause is too complex for.

```
It only tells Meteor that the package is being treated like an NPM package.
```

Why you want a package can be treat like a NPM package?

**Comment by trusktr** [comment]
> But still we need a way to define npm modules/packages for import it in a isomorphic way, say on client, server, or both. 

This already works. Using `api.usesNpmModules` (or whatever it is called), you put the rest in `package.json`, then you can make `server/main.js` or `client/main.js` and those entry points will only load in server or client, respectively. If you have `main.js` in the root of the package, that can be for both. In `package.js` you can do as normal:

```js
api.addFiles('client/main.js', 'client')
api.addFiles('server/main.js', 'server')
api.addFiles('main.js') // both
// etc
```

**Comment by trusktr** [comment]
@benjamn @avital @stubailo Any feedback on this? I think the `package.js` API already has everything we need, except for a single function call. I used the wrong prefix in the above examples, it would be more like `Package.usesNpmModules` (or similar) then the rest would go in `package.json`, and as described in my previous comment, you can simply specify which target the files are for using the existing `api.addFiles` API. I don't really see why any more is needed.

**Comment by avital** [comment]
I'm having a hard time following this thread, or even understand the request.

Is it "packages should be able to require NPM modules installed in the app's top-level `node_modules` directory?"

**Comment by bySabi** [comment]
yes it is!
package must require NPM´s on app's top-level `node_modules`. This needed among others things for properly support local NPM modules.

**Comment by avital** [comment]
Our plan is to just make that work -- so that `require('foo')` from within any package will also look in the `/node_modules/foo` directory in your app. But you won't be able to require specific `.js` files from the top level of your app.

For example, this will probably be the future best practice for using React -- you'll run `npm install react` at the top-level of your app, and any React-related packages will `require('react')`, printing an explanation if it can't find the module in your app.

Does this resolve the issue without requiring any new APIs?

**Comment by bySabi** [comment]
`require('foo')` from app `node_modules` is enough. And as you say import/require `.js` file outside of packages is not needed neither recommended.

MDG plans for  `Npm.depends` include search for modules first on app `node_modules` and install it if not found? What will happend with packages relay on npm modules installed on `.npm/node_modules`?

We still need express npm modules dependencies in `package.js` and now for client and server code.

**Comment by avital** [comment]
Are you asking: "If a user installs package foobar via `Npm.depends` in their package and via `npm install` in their app, and you call `require('foobar')` in that package, which will it find?". Not sure, what do you think the right answer is?

**Comment by bySabi** [comment]
Meteor packages should resolve npm modules dependecy similar to npm3: https://docs.npmjs.com/how-npm-works/npm3:
  - Search on top `node_modules` and install it if not there.
  - If the module version found mismatch then install it on `meteor-foo-package\.npm\node_modules`

I start this thread cause I make some local npm modules and put on `packages` folder. Something like this:

```
my_meteor_app
  client
  server
  packages
    meteor-package-ex1
    npm-package-ex1
    npm-package-ex2
```

and install it with:
```
  npm install packages/npm-package-ex1 --save
```

But I can´t import `npm-package-ex1` inside `meteor-package-ex1`. Then I was trying figure out how should we add local npm dependecies to packages `package.js`.

**Comment by bySabi** [comment]
@trusktr think that we should put a `package.json` settings file on new 1.3 packages that require npm modules just like we did at app top level and I agree him.
But I think too,  is easy for incomers a sintax like this: 
```
...
  api.use([
    {                        // npm packages
      "react": "^0.14",
      "react-redux": "^0.1.1"
    },
    'react-tab@0.15.0'       // a meteor package
  ], client);
...
``` 

Then we spinning at this point :-)

**Comment by trusktr** [comment]
> Our plan is to just make that work -- so that require('foo') from within any package will also look in the /node_modules/foo directory in your app.

How will the packages be installed there? Automatically, or the developer has to specify them? If the app developer has to specify it, this will overload the app developer with extra work that isn't normally needed in other setups that use NPM. The app dev shouldn't necessarily need to worry about which NPM dependencies a Meteor-package needs, in the same vein as current NPM users don't have to manually specify dependencies of dependencies. So, if Meteor will detect and install dependencies (into the top-level if that what you guys want to do) that's fine, as long as it's automatic, based only on

1. The NPM dependencies the app dev explicitly needs, listed in package.json
2. The Meteor-package dependencies the app dev explicitly needs, listed in .meteor/packages

**Comment by avital** [comment]
While I agree that requiring the app developer to manually run `npm install` is suboptimal, I think the difference in user experience is actually minimal. Compare this to all of the overhead people go through building "traditional" Node apps. While Meteor aims to have a better developer experience than Node, I think we need to free ourselves of our self-imposed shackles of "everything should work 100% as easily as remotely imaginable".

The solution I propose is the following convention: Imagine a package, say, `react-meteor-data` needs the `react` module installed in the app. The package will try calling `require('react')`. If that fails, it will print the following message in the server console or a overlay message on the page: "Please run `npm install react --save` in your app directory".

This approach even has an added benefit: App developers now know that they have complete control over which version of React they use, without any "Meteor magic".

To me, that sounds like a 99% optimal solution that doesn't require additional machinery in the meteor tool. Any additional machinery requires design, may have bugs and incurs ongoing maintenance costs. In short, more machinery means slower progress on the Meteor framework. Slow progress on Meteor seems like a bigger problem than "apps with react need to run one additional step that's explicitly printed out for them".

The longer term plan for Meteor is to have all packages just be NPM packages. In that world, this problem is solved by having the `react-meteor-data` NPM package will depend on the `react` NPM package, which will be installed at the top-level of your app due to NPM3 deduplication.

**Comment by bySabi** [comment]
Supposedly 'machinery' could be add in the future!

Anything to the version `1.3` is released soon.

**Comment by casperan** [comment]
avital, that's a sane approach, great.
if anything, any machinery/magic should only/could be added on top by a
package.

**Comment by trusktr** [comment]
> "Please run `npm install react --save` in your app directory".

What if a package needs to `require('something')`, and that single require call ends up needing to `require` 300 other things. Will Meteor just print 300 of those, [like this](http://trusktr.io:7777/cinuhepogo)?

> App developers now know that they have complete control over which version of React they use, without any "Meteor magic".

Without dependency management? It might make sense to have a singleton instance of React, but not necessarily for other libraries. Take underscore for example. If the API ever breaks there ([which is has, without semantic changes to it's version number](https://gist.github.com/jashkenas/cbd2b088e20279ae2c8e)), you'll have app developers and packages possibly wanting multiple differently-behaving versions of underscore, which is perfectly functional (unlike React's singleton requirement). Thus, a package might need some version of underscore like `<X.X.X`  while some other package or the app needs the incompatible `>X.X.X` version. How will Meteor handle those cases (dependency forks)?

> The longer term plan for Meteor is to have all packages just be NPM packages.

Ah, well, in that case, then yeah, it'd be solved. :]

**Comment by avital** [comment]
@trusktr You probably don't want to send down multiple versions of underscore to the client (due to inflated bundle size), so what you call a feature I call a bug. If you want your package to work with a specific version of underscore rather than using a single version used in your app, just do that yourself by running `npm install underscore` in your package directory.

And I don't understand your first point -- If the first require suggests you install a module that should be enough. To clarify, my suggestion is not that Meteor does anything special; it's that packages that want to work with a single install of a module at the top-level of your app write the code, e.g.:

```
var foo;
try {
  foo = require('bar');
} catch (e) {
  console.error("Please run ...");
  process.exit(1);
}  
```

**Comment by trusktr** [comment]
> You probably don't want to send down multiple versions of underscore to the client

I would want to avoid it, definitely. It was only one example. I've definitely had cases where I had to though. It's often a lot more work to fix that dependency problem than the overhead of just bundling (the parts of) two (es6 statically trimmed, tree shaken) libraries into your app.

> If the first require suggests you install a module that should be enough.

You'd be surprised: people can do strange things with modules. For example, they might install module A, with dependency chain `A -> B -> C -> D -> D`, and then in their entry point, the first thing that is imported is

```
var d = require('D')
```

Would Meteor still tell the user to install `A` in that case?

> ```
> var foo;
> try {
>   foo = require('bar');
> } catch (e) {
>   console.error("Please run ...");
>   process.exit(1);
> }  
> ```

That's too much extra work that we have to account for modules on NPM not already doing. Plus, what if you install 10 packages that all do that, all specifying different versions of the dependency? There might possibly be a dependency fork requirement. 

I definitely agree we should avoid dependency forks, but sometimes it's more convenient not to (especially if a tool like `Webpack` (and now Meteor 1.3?) will bundle only the things of a library that are imported, so that if we depend on two versions of a library that has hundreds exports we might only end up with two of them in our app if we've only ever imported one from each.

**Comment by trusktr** [comment]
By the way, NPM 3.x handles dependency forks by installing those in a module's `node_modules` folder instead of the top level. If there are no forks, then everything goes at the top level, flat.

**Comment by mbabauer** [comment]
This was a very interesting discussion to read, but did anything ever become of this?  I am in a similar situation.  I am developing a set of apps that will have  a shared Meteor package that uses a custom NPM module.  My folder structure looks like:

myApp1
    packages
        myMeteorPackage1 (softlink)
myApp2
    packages
        myMeteorPackage1 (softlink)
myMeteorPackage1
myNpmPackage

I have ran `npm link ../myNpmPackage` in myMeteorPackage1, and then created a soft link to my shared package, but this doesn't seem to work.  I also tried creating a `node_modules` under my package and soft-linking to my NPM package, but that didn't work either.

**Comment by trusktr** [comment]
@mbabauer Your NPM package needs to be inside of node_modules of each app. I would suggest for you to learn the `npm link` workflow to link your NPM package

**Comment by mbabauer** [comment]
I looks like that was the part I was missing:  Including the NPM module in my actual project.  I was kinda hoping that including it in the Meteor plugin would include it by proxy, but I guess that's not how it works.  Thanks for the assist, though.

**Comment by stubailo** [comment]
>  was kinda hoping that including it in the Meteor plugin would include it by proxy, but I guess that's not how it works.

We thought about a million different ways of doing this, and none of them seemed like a good idea. Managing dependencies across different package systems is likely to become a nightmare.

**Comment by mbabauer** [comment]
Yeah, that makes sense.  Can't tell you how many sleepless nights I have had with Java dependencies in Maven, where one lib requires a version not supported by another.

**Comment by nicolaslopezj** [comment]
Meteor packages should have a peerDependecies option for node packages. I think that will solve this problem.

**Comment by djhutch** [comment]
@avital: we are using react with meteor. we have most of our code and tests in packages and use practicalmeteor:mocha. Previously we used Npm.depends in our package to include the react dependencies. When we moved to meteor 1.3 (currently beta.16) we used npm install to add our react dependencies to the app and removed the Npm.depends call in our package. Unfortunately this appears to break test-packages. We get an error that the react module cannot be loaded. If we add the Npm.depends back to our package but also keep react in our app’s npm dependencies then test-packages works, but our react ‘refs’ fail in the app because react appears to be loaded multiple times. Should test-packages still work without Npm.depends calls in 1.3? What is the proper way to add npm dependencies for our packages in 1.3? Thanks!

**Comment by abernix** [comment]
Closing this because I believe you'll find that NPM (published and non-published) work fine with the final Meteor 1.3.  For non-published packages you should read about [`npm link`-ing your modules appropriately](https://docs.npmjs.com/cli/link).  Please read about [using NPM packages with Meteor](http://guide.meteor.com/using-npm-packages.html) in the Guide.

Post any questions on the Meteor forums or if you find bugs, please open issues with reproductions of your problem.
